cheatcodes/src/cheatcodes.cairo

```diff
@@ -34,54 +44,74 @@ impl RevertedTransactionImpl of RevertedTransactionTrait {
     }
 }
 
-fn declare(contract: felt252) -> ClassHash {
-    let span = cheatcode::<'declare'>(array![contract].span());
+impl ContractClassImpl of ContractClassTrait {
+    fn precalculate_address(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> ContractAddress {
+        let class_hash: felt252 = (*self.class_hash).into();
+        let mut inputs: Array::<felt252> = _prepare_calldata(class_hash, constructor_calldata);
 
-    let exit_code = *span[0];
-    let result = *span[1];
-    assert(exit_code == 0, 'declare should never fail');
-    result.try_into().unwrap()
-}
+        let outputs = cheatcode::<'precalculate_address'>(inputs.span());
+        (*outputs[1]).try_into().unwrap()
+    }
 
-fn deploy(prepared_contract: PreparedContract) -> Result::<ContractAddress, RevertedTransaction> {
-    let PreparedContract{class_hash, constructor_calldata } = prepared_contract;
-    let mut inputs = array![class_hash.into()];
+    fn deploy(self: @ContractClass, constructor_calldata: @Array::<felt252>) -> Result<ContractAddress, RevertedTransaction> {
+        let class_hash: felt252 = (*self.class_hash).into();
```

Comment:
I think you can perform conversion inside of `_prepare_calldata` function instead

Reply:
```
fn _prepare_calldata(class_hash: @ClassHash, constructor_calldata: @Array::<felt252>) -> Array::<felt252>  {
    let class_hash: felt252 = (class_hash.clone()).into();
```
but I have to clone class_hash. Is it ok? @Arcticae